### PR TITLE
fix dangling pointer to state buffer in state registry

### PR DESCRIPTION
**Pull request by silvansievers**

### CODE REVIEW

**Review by FlorianPommerening — COMMENTED** [review]
I'm fine with the pull request as is but would prefer some minor changes. Also, since this is such a central part of the code, I'd say it needs an experiment, maybe with blind search, to see if the performance is affected. Although, we don't have much choice since we have to fix the bug.

Independent of the issue, we could use this as a starting point to discuss the refactoring that was planned anyway, where we move successor generation out of the registry. Maybe a topic for a meeting?

File: src/search/state_registry.cc [hunk 1]
```diff
@@ -91,6 +93,17 @@ State StateRegistry::get_successor_state(const State &predecessor, const Operato
...
         StateID id = insert_id_or_pop_state();
+        /*
+          insert_id_or_pop_state deletes the state data pointed to by buffer
+          if the state already exists. We need to use the buffer corresponding
+          to the right state. NOTE: the below code is exactly what lookup_state
+          does, but for symmetry with the above case for axioms and for future
+          refactoring, we explicitly set buffer here.
+
+          TODO: ideally, we would not modify state_data_pool here and in
+          insert_id_or_pop_state, but only at one place.
```
[the remark this answers is not in the document]
Comment (silvansievers) [reply src/search/state_registry.cc:104]: We are: `get_successor_state` pushes something into `state_data_pool` and a different method, namely `insert_id_or_pop_state` possibly pops that something from `state_data_pool` again. I think this lead to this error in the first place. Having only one method modify `insert_id_or_pop_state` would avoid that error likely.

File: src/search/state_registry.cc [hunk 2]
```diff
@@ -91,6 +93,17 @@ State StateRegistry::get_successor_state(const State &predecessor, const Operato
             }
         }
         StateID id = insert_id_or_pop_state();
+        /*
+          insert_id_or_pop_state deletes the state data pointed to by buffer
+          if the state already exists. We need to use the buffer corresponding
+          to the right state. NOTE: the below code is exactly what lookup_state
+          does, but for symmetry with the above case for axioms and for future
+          refactoring, we explicitly set buffer here.
```
[the remark this answers is not in the document]
Comment (silvansievers) [reply src/search/state_registry.cc:101]: Also `TaskProxy::create_state` is in the public interface. I think overloading `lookup_state` makes some sense, even though I also don't really like it.

[the remark this answers is not in the document]
Comment (FlorianPommerening) [reply src/search/state_registry.cc:101]: It's not great to do it in `TaskProxy` either. I think we have a comment there that the method is public but should only be used by the registry. This whole part probably needs an overhaul at some point, where we move the state creation out of the registry entirely. For now, I'd keep the axiom case above as is, and only call `lookup_state` here.

File: src/search/state_registry.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (FlorianPommerening) [reply src/search/state_registry.cc:104]: OK, I get what you mean. I thought "here" referred to the line below the comment. Maybe move the comment to the place where the modification happens, or explain why we need to re-set `buffer` here with a reference to the issue.

File: src/search/state_registry.cc [hunk 3]
```diff
@@ -64,6 +70,11 @@ const State &StateRegistry::get_initial_state() {
 //     operating on state buffers (PackedStateBin *).
 State StateRegistry::get_successor_state(const State &predecessor, const OperatorProxy &op) {
     assert(!op.is_axiom());
+    /*
+      TODO: ideally, we would not modify state_data_pool here and in
+      insert_id_or_pop_state, but only at one place, to avoid errors like
+      buffer becoming a dangling pointer.
```
Comment (FlorianPommerening) [inline src/search/state_registry.cc:76]: I'd add a reference to the issue here.

**Review by FlorianPommerening — APPROVED** [review]
I'm still fine with merging, as long as the experiments do not show anything to worry about.

**Pull request closed without merging**